pkg/operator/controller/dnsnameresolver-crd/controller.go

```diff
@@ -0,0 +1,150 @@
+)
+
+const (
+	controllerName = "dnsnameresolver_crd_controller"
+)
+
+// New creates and returns a controller that creates DNSNameResolver CRD when the
+// appropriate featuregate is enabled.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		cache:  mgr.GetCache(),
+		client: mgr.GetClient(),
+		config: config,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	clusterNamePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		expectedName := operatorcontroller.FeatureGateClusterConfigName()
+		actualName := types.NamespacedName{
+			Namespace: o.GetNamespace(),
+			Name:      o.GetName(),
+		}
+		return expectedName == actualName
+	})
+	if err := c.Watch(source.Kind(mgr.GetCache(), &configv1.FeatureGate{}), &handler.EnqueueRequestForObject{}, clusterNamePredicate); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DNSNameResolverEnabled indicates that the "DNSNameResolver" featuregate is enabled.
+	DNSNameResolverEnabled bool
+	// DependentCaches is a list of caches that are used by Controllers watching DNSNameResolver
+	// resources. The dnsnameresolver_crd controller starts these caches once
+	// the DNSNameResolver CRD has been created.
+	DependentCaches []cache.Cache
+	// DependentControllers is a list of controllers that watch DNSNameResolver
+	// resources. The dnsnameresolver_crd controller starts these controllers once
+	// the DNSNameResolver CRD has been created and the DependentCaches are started.
+	DependentControllers []controller.Controller
+}
+
+// reconciler handles the actual featuregate reconciliation logic in response to
+// events.
+type reconciler struct {
+	config Config
+
+	cache            cache.Cache
+	client           client.Client
+	startCaches      sync.Once
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a FeatureGate and creates or
+// reconciles the DNSNameResolver CRD.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logrus.Info("reconciling", "request", request)
+
+	// Check if the DNSNameResolved feature gate is enabled or not.
+	if !r.config.DNSNameResolverEnabled {
+		return reconcile.Result{}, nil
+	}
+
+	// Get the feature gate object.
+	featureGate := configv1.FeatureGate{}
+	if err := r.cache.Get(ctx, request.NamespacedName, &featureGate); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the feature set is either TechPreviewNoUpgrade or CustomNoUpgrade.
+	if featureGate.Spec.FeatureSet != configv1.TechPreviewNoUpgrade && featureGate.Spec.FeatureSet != configv1.CustomNoUpgrade {
+		return reconcile.Result{}, nil
+	}
```

Comment:
You shouldn't need all this featuregate logic in `Reconcile`; the `if !r.config.DNSNameResolverEnabled {` check should be sufficient.  If you were following an example set in in an older version of cluster-ingress-operator, please refer to https://github.com/openshift/cluster-ingress-operator/pull/908/commits/34dc8bd5712c584dfbe27e686f805bba7ff17ed8 which shows how we simplified the featuregate logic using library-go in cluster-ingress-operator.

Reply:
Oh, I see—this is more about checking _which one_ of TechPreviewNoUpgrade or CustomNoUpgrade is enabled, so that you know which version of the CRD to create.  I'm surprised the consumer of the FeatureGate API has to check this.  If this is correct behavior, I believe it should be codified in [library-go's `FeatureGateAccess` controller](https://github.com/openshift/library-go/blob/126b47137408f93a2fc9f2e8cbeedda61f9ebbfb/pkg/operator/configobserver/featuregates/simple_featuregate_reader.go#L26).  We should check with @deads2k.

Reply:
It was pointed out by @JoelSpeed during the review of the DNSNameResolver api [PR](https://github.com/openshift/api/pull/1524#discussion_r1262678270) that adding the CustomNoUpgrade CRD is also required without which the CustomNoUpgrade uses cases will be broken. Maybe, @JoelSpeed, you can point to some existing best practices of how to enable either type of the CRD.

Reply:
You can use the CVO to install the CRD, it understands the annotations and will install the correct version of the CRD depending on the feature set that the cluster is installed as